### ISSUE TITLE
Rebuild database (and re-populate a borked homepage that lacks all downloads) in one click?

### ISSUE THREAD

**Issue by GlassedSilver** [Tzahi12345/YoutubeDL-Material#210]
Blocked by: #570 

<hr>

Hey! So...

Unfortunately I had a major mishap with my unRAID server and whilst all downloads and all the info jsons and such are available, it would be nice not to re-download everything. My database file got yeeted, so...

Yes, I'm actually staring at the "fresh install" experience.

We could scan the folder structure? Or maybe something else... I don't know. This might be related to #157, but I think it's still fairly different.

Luckily I was still in the evaluation phase of sorts and just slowly getting started more ever since some important fixes had been implemented in nightly.

*sigh*

What a weekend... At least life's good in other ways and basically nothing but meta data for other apps had been lost. Still sucks a lot, since that can be harder to repopulate than the data itself, but... Here we are. :D

**Comment by Tzahi12345** [comment]
Sorry about the data loss! It's always quite painful.

There's actually a mechanism to fix this in place, but it won't do it from one click. This issue/ticket can address that specifically, but I'll help you get your data back in the meantime.

**Background:** the way it works now is like you suggested, each folder is scanned and if a video missing from its proper DB location, it gets individually added to the DB.

The process goes like this: get all users->check each of their `video` and `audio` folders, as well as their subscription directories->add missing videos if caught.

Your videos are not getting re-added because when it tries to get a list of users (for a list of folders to check), it comes up on an empty list.

**Fix:** if we recreate your user in the DB, your user's folder will be checked on startup. This is great, because it contains all your videos! On startup they will be imported.

1. Recreate the admin account from the login screen, and on server restart that account will get its videos back. You have to manually create each account from the UI (shouldn't take too long) and those user's videos will get re-added.

2. Because your subscriptions are also missing from the db (new accounts have no subscriptions), you'll have to do the same for each user's subscriptions. Remember to resubscribe with the same settings (name and audio only must match), and then those folders will also be checked on startup.

If you're not using multi-user mode, you can skip to step 2. It's not the most convenient process, but I just tested it out and it could get you back on your feet in minutes (instead of hours or days). Let me know how it goes or if you need any help throughout!

**Comment by GlassedSilver** [comment]
Interesting... Well coincidentally I'm actually in a non-multi-user environment and nothing shows up. I had not set up subscriptions so far, because I didn't want old issues becoming major future worries like the limited arg and quality support.

Now that I'm all ready and fired up to go... this happens. XD

Either way, learnt my lesson... CA app data backup separately will be set up. (I already have a hunch what happened, doubt it's gonna happen again, but there's so many ways to mess something up :P)

**Comment by Tzahi12345** [comment]
>Interesting... Well coincidentally I'm actually in a non-multi-user environment and nothing shows up.

Well, this should make things quite easy then, especially since you don't have any subscriptions. Are you running on the latest `nightly`? If you are, and there are videos in your video and audio directory, they should show right back up in the DB. Maybe there's an error in your log? Set the logging to verbose if you haven't.

**Comment by GlassedSilver** [comment]
> 
> 
> > Interesting... Well coincidentally I'm actually in a non-multi-user environment and nothing shows up.
> 
> Well, this should make things quite easy then, especially since you don't have any subscriptions. Are you running on the latest `nightly`? If you are, and there are videos in your video and audio directory, they should show right back up in the DB. Maybe there's an error in your log? Set the logging to verbose if you haven't.

I download to my own directory logic btw, maybe that's the problem?

`video/`  custom output: `%(uploader)s [%(channel_id)s]/%(upload_date)s - %(title)s [%(id)s]` + extension then of course

It is very important to me to keep it in that logic.

**Comment by Tzahi12345** [comment]
@GlassedSilver Just pushed a fix that should allow files in subfolders to be autorecognized -- there was a bug in the logic that was fixed [here](https://github.com/Tzahi12345/YoutubeDL-Material/commit/46ed0fe992f5b3cf002fc12d20e1283417f2d76e). Try pulling the latest nightly and restarting the server, your videos should all be there!

**Comment by GlassedSilver** [comment]
Still coming up empty...

**Comment by Tzahi12345** [comment]
Oops, there was another bug hidden there. This time, UNIX paths (with / instead of \\\\) were getting improperly parsed during the registration process (`PewDiePie/somevideo.mp4` became `ewDiePie/somevideo.mp4`). Anyways, I've gone ahead and pushed a fix -- let me know if it works! (it really should this time lol)

**Comment by GlassedSilver** [comment]
Works! Thanks a ton!

One problem though is that I have (and have been having before this) many thumnail-less videos.

I've always custom-arged to download the thumbnail as an image file, would it be possible to populate the thumbnail for the db from the filesystem instead of just pulling at download time? (iirc we are now pulling the thumbnail into the db at download time, right? Or was that just planned so far?)

Either way, thanks a ton. This is a BIG help.

**Comment by Tzahi12345** [comment]
No problem!

We recently changed up how thumbnails work. If they are downloaded along with a video (this is done by default), the thumbnail path will be stored in the DB ([code reference](https://github.com/Tzahi12345/YoutubeDL-Material/blob/5eccaa13e5cc5299fab707210d72b2fbabedccbf/backend/db.js#L30)). When a user requests the list of recent videos, each video will be checked for a thumbnailPath. If it exists and if the thumbnail actually exists (it might have been deleted), then the thumbnail blob (i.e. the actual thumbnail) will be sent over to the UI along with other metadata.

The reason why I mention this is because if the thumbnailBlob is not received by the UI, the thumbnailURL will then be used. The URL is retrieved from youtube-dl, but it's not necessary valid for all time. With the new system, the thumbnail is downloaded by default and this should solve that problem, in that there will be a local cache of the thumbnail which could be used.

I'm assuming your old videos are the ones with a missing thumbnail. The only fix for this would be redownloading the video -- alternatively you could manually download just the thumbnail and put in a thumbnailPath for all your videos. The latter is super tedious, but could be easily automated. Maybe a dialog to import unregistered files (meaning deleted from the DB) could include an option to fix thumbnails. And within that dialog we can also have options like "recover users and subscriptions" which will rebuild the database as best we can.

**Comment by GlassedSilver** [comment]
Ah thanks a lot for the insight!

Yeah the repair function as you describe it could be helpful indeed, because especially in the beginning I concentrated on videos I expected to get yeeted. (I was right about a good amount of them)

**Comment by Atreidae** [comment]
Just a suggestion. Should this get thrown into the wiki?
I'm looking at importing files from another YTDL instance as several channels I follow have been deleted prior to me starting to back them up and it would have been nice to see it in there.

**Comment by Tzahi12345** [comment]
Here's a little thought I had:

We can locally store a JSON in each user directory and each subscription directory (similar to the `info.json` for files). We can then easily rebuild the database as we don't have to make any guesses regarding users or subscriptions.
